### ISSUE TITLE
'i' is defined but never used.

### ISSUE THREAD

**Issue by ficristo** [jshint/jshint#2784]
The following code pass with 2.8.0 but will fail with jshint 2.9.1-rc1 with:
`test.js: line 4, col 14, 'i' is defined but never used.`
To me seems a bug. (I'm not sure if it is a duplicate or not)

```javascript
/*jshint unused:vars, -W004*/
function test() {
    var array = ['a', 'b', 'c'];
    for (var i in array) {
        var elem = array[i];
        elem[i] = 0;

        var i = "test";
        i = 0;
    }
}
test();
```

**Comment by jugglinmike** [comment]
Thanks for checking out the release candidate and reporting back! Although this
is a new error for your code, I believe it is actually a bug fix.

Historically, JSHint has treated `var` bindings as block-scoped despite the
fact that they are function-scoped. It exposes an option dedicated to disabling
this (technically incorrect) behavior, [`funcscope`](). From the docs:

> This option suppresses warnings about declaring variables inside of control
> structures while accessing them later from the outside. Even though
> JavaScript has only two real scopes—global and function—such practice leads
> to confusion among people new to the language and hard-to-debug bugs. This is
> why, by default, JSHint warns about variables that are used outside of their
> intended scope.
>
>     function test() {
>       if (true) {
>         var x = 0;
>       }
>
>       x += 1; // Default: 'x' used out of scope.
>                 // No warning when funcscope:true
>     }

In keeping with that assumption, the first declaration of `i` in your code
should be scoped to [the special block reserved for the `for` statement's
head](http://www.ecma-international.org/ecma-262/6.0/#sec-runtime-semantics-forin-div-ofheadevaluation-tdznames-expr-iterationkind).
The second declaration of `i` shadows the first, and all further references are
to that second binding.

Here's a reduced test case:

```js
// jshint unused: true, shadow: true
(function() {
  var x;
  {
    var x;
    void x;
  }
}());
```

For consistency's sake (and even though it is a new error), JSHint should
consider the `var` from line 3 as unused.

But I don't think we're off the hook completely--if `funcscope` is really intended to control the way JSHint interprets `var` bindings, then it ought to effect more than just shadowing detection. [I've opened a pull request](https://github.com/jshint/jshint/pull/2789) to ensure that *this* code does not trigger warnings:

```js
// jshint funcscope: true, unused: true, shadow: true
(function() {
  var x;
  {
    var x;
    void x;
  }
}());
```

If that lands, then you will just need to set `funcscope: true` in your JSHint
configuration. It's a little controversial, though, so I'm looking for feedback
from you and the JSHint maintainers!

**Comment by ficristo** [comment]
The code I written in the first comment come from a project where we introduced jshint only after a while. In that time we written some suboptimal code like that (very simplified) snippet.

I tried jshint on master and with the PR: adding `funcscope:true` resulted in an error on master and no error with the PR.
So to me is fixed. Thank you @jugglinmike

**Comment by jugglinmike** [comment]
After discussing this with the project maintainers, I decided that reliance on `funcscope` is not the best way forward. (In some future version of JSHint, we'd like `var` bindings to follow function-scoping semantics by default, and optionally allow users to opt-in to *complete* application of ES2015 block scoping.)

It's true that the code you shared is a little gnarly (and probably uncommon), but I was able to find a solution that preserves existing behavior; see gh-2795. If *that* patch lands, then you won't need to change your configuration at all.
